### PR TITLE
makefile: workaround boost errors under latest compilers

### LINKED ISSUE #300: address warnings in resource/resource_gen.cpp etc, then revert workaround
As discussed in #298, g++ 5.4.0 issues some new warnings that must be disabled in order to be non-fatal due to `-Werror`.  In order to support this compiler, several `-Wno-*` options were added to resources/Makefile.am (8a4b016c4b39938cf8a3041d8b006b163d02c3cc).   However, one that is needed, `-Wno-maybe-uninitialized`, could not be added because it is not understood by clang 3.8 in travis, and an unknown warning option is fatal.

The workaround (68569905a21592f290bf581a78e270336df450cd) was to add `-Wno-error` to resources/Makefile.am, but that should only be a temporary solution until the root cause of the warnings is found.

Here is the warning output:
```
jimbo:~/proj/flux-sched/resource > make
Making all in planner
make[1]: Entering directory '/home/garlick/proj/flux-sched/resource/planner'
Making all in .
make[2]: Entering directory '/home/garlick/proj/flux-sched/resource/planner'
  CC       libplanner_la-planner.lo
  CCLD     libplanner.la
ar: `u' modifier ignored since `D' is the default (see `U')
make[2]: Leaving directory '/home/garlick/proj/flux-sched/resource/planner'
Making all in test
make[2]: Entering directory '/home/garlick/proj/flux-sched/resource/planner/test'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/home/garlick/proj/flux-sched/resource/planner/test'
make[1]: Leaving directory '/home/garlick/proj/flux-sched/resource/planner'
Making all in .
make[1]: Entering directory '/home/garlick/proj/flux-sched/resource'
  CXX      utilities/utilities_grug2dot-grug2dot.o
  CXX      utilities_grug2dot-resource_gen_spec.o
In file included from /opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:33:0,
                 from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp: In member function ‘void boost::vec_adj_list_impl<Graph, Config, Base>::copy_impl(const boost::vec_adj_list_impl<Graph, Config, Base>&) [with Graph = boost::adjacency_li/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp:95:84: warning: ‘*((void*)& ei_end +32)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
               && edges BOOST_GRAPH_MEMBER first != x.edges BOOST_GRAPH_MEMBER first);
                                                                                    ^
In file included from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246:0,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:2119:27: note: ‘*((void*)& ei_end +32)’ was declared here
         edge_iterator ei, ei_end;
                           ^
In file included from /opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:33:0,
                 from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp:70:9: warning: ‘*((void*)& ei +48)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (edges BOOST_GRAPH_MEMBER first == edges BOOST_GRAPH_MEMBER second) 
         ^
In file included from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246:0,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:2119:23: note: ‘*((void*)& ei +48)’ was declared here
         edge_iterator ei, ei_end;
                       ^
  CXXLD    utilities/grug2dot
  CXX      utilities/utilities_resource_query-resource-query.o
  CXX      utilities/utilities_resource_query-command.o
  CXX      utilities_resource_query-dfu_traverse.o
  CXX      utilities_resource_query-dfu_traverse_impl.o
  CXX      utilities_resource_query-resource_gen.o
In file included from /opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:33:0,
                 from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246,
                 from resource_graph.hpp:29,
                 from resource_gen.hpp:30,
                 from resource_gen.cpp:30:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp: In member function ‘void boost::vec_adj_list_impl<Graph, Config, Base>::copy_impl(const boost::vec_adj_list_impl<Graph, Config, Base>&) [with Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>; Config = boost::detail::adj_list_gen<boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>, boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t, boost::no_property, boost::listS>::config; Base = boost::directed_graph_helper<boost::detail::adj_list_gen<boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>, boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t, boost::no_property, boost::listS>::config>]’:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp:70:9: warning: ‘*((void*)& ei +48)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (edges BOOST_GRAPH_MEMBER first == edges BOOST_GRAPH_MEMBER second) 
         ^
In file included from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246:0,
                 from resource_graph.hpp:29,
                 from resource_gen.hpp:30,
                 from resource_gen.cpp:30:
/opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:2119:23: note: ‘*((void*)& ei +48)’ was declared here
         edge_iterator ei, ei_end;
                       ^
  CXX      utilities_resource_query-resource_gen_spec.o
In file included from /opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:33:0,
                 from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp: In member function ‘void boost::vec_adj_list_impl<Graph, Config, Base>::copy_impl(const boost::vec_adj_list_impl<Graph, Config, Base>&) [with Graph = boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>; Config = boost::detail::adj_list_gen<boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>, boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t, boost::no_property, boost::listS>::config; Base = boost::directed_graph_helper<boost::detail::adj_list_gen<boost::adjacency_list<boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t>, boost::vecS, boost::vecS, boost::directedS, Flux::resource_model::resource_pool_gen_t, Flux::resource_model::relation_gen_t, boost::no_property, boost::listS>::config>]’:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp:95:84: warning: ‘*((void*)& ei_end +32)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
               && edges BOOST_GRAPH_MEMBER first != x.edges BOOST_GRAPH_MEMBER first);
                                                                                    ^
In file included from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246:0,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:2119:27: note: ‘*((void*)& ei_end +32)’ was declared here
         edge_iterator ei, ei_end;
                           ^
In file included from /opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:33:0,
                 from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adj_list_edge_iterator.hpp:70:9: warning: ‘*((void*)& ei +48)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (edges BOOST_GRAPH_MEMBER first == edges BOOST_GRAPH_MEMBER second) 
         ^
In file included from /opt/boost-1.53/include/boost/graph/adjacency_list.hpp:246:0,
                 from resource_gen_spec.cpp:26:
/opt/boost-1.53/include/boost/graph/detail/adjacency_list.hpp:2119:23: note: ‘*((void*)& ei +48)’ was declared here
         edge_iterator ei, ei_end;
                       ^
  CXXLD    utilities/resource-query
make[1]: Leaving directory '/home/garlick/proj/flux-sched/resource'
```

### LINKED ISSUE #732: Issues building on linux-rhel7-power9le with Spack (GCC 10.2.0)
This is a followup to flux-framework/flux-core#3133 ; after resolving that issue (which occurred while building `flux-core`), I immediately ran into further problems building `flux-sched@master+cuda`. As with the prior linked issue, I'm using GCC 10.2.0, installed via Spack, on a `linux-rhel7-power9le` system.

The build errors I'm getting this time seem to be within templated C++ code, so I'm not going to copy the error text in the issue body; instead, I have the Spack build log here: [spack-build-out.txt](https://github.com/flux-framework/flux-sched/files/5106461/spack-build-out.txt)

The exact spec that is failing to build is `flux-sched@master%gcc@10.2.0+cuda`.

### LINKED ISSUE #747: sched-simple issues with quick start on 0.8.0 release
Trying to follow the quick start guide. In the process of working through issues on a CTS platform (either spack is providing a dirty pkgconf environment or the flux-core and flux-sched autotools setup is reading things incorrectly and I need to debug that) but was able to get a completed install on a dev cluster as per [](https://flux-framework.readthedocs.io/en/latest/quickstart.html#building-the-code )on a fairly recent spack hash with

```bash
spack install flux-sched@0.8.0%gcc@9.3.0
# flux-core@0.16.0%gcc@9.3.0~cuda~docs arch=linux-centos7-haswell
# flux-sched@0.8.0%gcc@9.3.0~cuda arch=linux-centos7-haswell
```

```flux keygen``` works but 

```bash
[rspavel@cn103 fluxTesting]$ flux start --size=4
2020-09-04T15:48:11.826345Z broker.err[0]: rc1: flux-module: cmb.rmmod sched-simple: No such file or directory
2020-09-04T15:48:11.827180Z broker.err[0]: Run level 1 Exited with non-zero status (rc=1) 2.7s
2020-09-04T15:48:12.190757Z broker.err[0]: rc3: flux-module: cmb.rmmod qmanager: No such file or directory
2020-09-04T15:48:12.226114Z broker.err[0]: rc3: flux-module: cmb.rmmod resource: No such file or directory
flux-start: 0 (pid 14789) exited with rc=1
```

when testing from within a slurm allocation.

It looks like this was a known issue as of #525 and a workaround exists. But I am unable to find said workaround and was instructed to make a new issue

### PR DESCRIPTION
**Pull request by SteVwonder**
Problem: when using the latest gcc/clang compilers, older boost versions
create multiple warnings to be emitted.  Our use of `-Werror` turns
these warnings in boost into hard errors, despite being in a third-party
dependency.

Solution: use `-isystem` rather than `-I` for the boost cppflags to
denote that the boost headers are system-provided headers, which turns
off most static analysis warnings for those headers.

Related Documentation:
- Clang: http://clang.llvm.org/docs/UsersManual.html#controlling-diagnostics-in-system-headers
- GCC: https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html

Closes #300 
Closes #747 
Closes #732

### CODE REVIEW

**Review by trws — APPROVED** [review]
This is how any system library like that should come in most likely, makes perfect sense.

**Comment by SteVwonder** [comment]
It looks like this workaround breaks when `BOOST_PREFIX=/usr`, as per this GCC bug report: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129

The two paths forward that I see are:

- Check if `BOOST_PREFIX` is one of the default system search paths, and if it is, still use `-I`.  Unfortunately, I think this means the warnings will only be silenced if boost isn't in a default system location
- Check if `BOOST_PREFIX` is one of the default system search paths, and if it is, set `BOOST_CPPFLAGS=""` and rely on the compiler picking up the header from one of the default paths
- Insert compiler-specific pragmas to turn off warnings for certain includes:
```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#pragma GCC diagnostic ignored "-Wunused-local-typedefs"
#include <third-party-library/header.h>
#pragma GCC diagnostic pop
```

**Comment by trws** [comment]
Default system paths are included by the `-isystem` method unless otherwise specified. If they are not explicitly added to a -I the warnings should be supppressed just by nature of coming from those places.  Note the test lower down in the bug where no listing and -I behave the same.

Sent from my iPad

On Sep 30, 2020, at 2:18 PM, Stephen Herbein <notifications@github.com> wrote:

﻿

It looks like this workaround breaks when BOOST_PREFIX=/usr, as per this GCC bug report: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70129

The two paths forward that I see are:

  *   Check if BOOST_PREFIX is one of the default system search paths, and if it is, still use -I. Unfortunately, I think this means the warnings will only be silenced if boost isn't in a default system location
  *   Check if BOOST_PREFIX is one of the default system search paths, and if it is, set BOOST_CPPFLAGS="" and rely on the compiler picking up the header from one of the default paths
  *   Insert compiler-specific pragmas to turn off warnings for certain includes:

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#pragma GCC diagnostic ignored "-Wunused-local-typedefs"
#include <third-party-library/header.h>
#pragma GCC diagnostic pop

**Comment by SteVwonder** [comment]
> Default system paths are included by the `-isystem` method unless otherwise specified. If they are not explicitly added to a -I the warnings should be supppressed just by nature of coming from those places.

Ok.  IIUC it sounds like `Check if BOOST_PREFIX is one of the default system search paths, and if it is, set BOOST_CPPFLAGS="" and rely on the compiler picking up the header from one of the default paths` is the way to go.  That also seems to be the way that CMake is handling this: https://gitlab.kitware.com/cmake/cmake/-/issues/16291

Any objections to just checking for `/usr/include` rather than going down the rabbit hole of querying each possible compiler for the full list of their default locations with a set of magic runes that is specific to each compiler? [GCC's magic runes, for example](https://stackoverflow.com/questions/4980819/what-are-the-gcc-default-include-directories).  According to the filed issues, it looks like `/usr/include` covers 99% of cases and the 1% edge cases are mostly cross-compilation.

**Comment by SteVwonder** [comment]
That latest change seems to have done the trick. Works on both the Travis tests and under Spack.  @trws , can you give the couple of lines at the end of the .am a quick look?

File: config/ax_boost_base.m4 [hunk 1]
```diff
@@ -299,4 +299,11 @@ AC_DEFUN([_AX_BOOST_BASE_RUNDETECT],[
     CPPFLAGS="$CPPFLAGS_SAVED"
     LDFLAGS="$LDFLAGS_SAVED"
 
+    dnl Due to a change in GCC 6+, searching a default system path via -isystem
+    dnl causes compilation issues.  Let the compiler find headers in default
+    dnl system paths "naturally".
+    if test "$BOOST_CPPFLAGS" = "-isystem /usr/include" ; then
```
Comment (milroy) [inline config/ax_boost_base.m4:305]: Is is possible that $BOOST_CPPFLAGS will resolve to something like `-isystem   __  /usr/include` (where `__` indicates multiple whitsepaces) or `-isystem /usr/include/`?  I don't see that it's defined elsewhere in flux-sched (and based on the file here these cases won't occur), but I want to make sure.

**Review by milroy — COMMENTED** [review]
LGTM!  This builds correctly on my test system.  I did ask one question, but I'm almost sure these changes can be merged.  Thanks!

File: config/ax_boost_base.m4 (hunk 1 above)
Comment (SteVwonder) [reply config/ax_boost_base.m4:305]: That's a good question.

>  I don't see that it's defined elsewhere in flux-sched (and based on the file here these cases won't occur), but I want to make sure.

That was the conclusion of my analysis as well.  The only other case I could think of was the user manually specifying the variable at the configuration line `BOOST_CPPFLAGS=/usr/include/ ./configure`, but I don't think we can really protect against the user shooting themselves in the foot in that scenario.

**Comment by SteVwonder** [comment]
Thanks @milroy and @trws! Setting the label so that we can tag 0.12.0 with this fix included.
